### PR TITLE
CorfuGuid: Retry if NTP is moving time ahead by crazy amounts

### PR DESCRIPTION
**Pull request by hisundar**
## Overview

Description:

Why should this be merged: 
Some systems the NTP is so misconfigured that time moves ahead by huge leaps only to get recorrected.
This causes some instability in the guid generator which relies on a sane NTP for performance.

Related issue(s) (if applicable): #2959 

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [ ] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/runtime/view/CorfuGuidGenerator.java [hunk 1]
```diff
@@ -244,12 +255,9 @@ private synchronized CorfuGuid getCorrectedTimestamp() {
                 }
                 resolutionCorrection = 0;
             }
-        } else {
-            resolutionCorrection = 0;
-        }
-
-        if (currentTimestamp < previousTimestamp) {
+        } else if (currentTimestamp < previousTimestamp) { // NTP corrects time backward
```
Comment (annym) [inline runtime/src/main/java/org/corfudb/runtime/view/CorfuGuidGenerator.java:258]: why do you need the condition? we already checked >, and =...

File: runtime/src/main/java/org/corfudb/runtime/view/CorfuGuidGenerator.java [hunk 2]
```diff
@@ -233,7 +234,17 @@ public long nextLong() {
      */
     private synchronized CorfuGuid getCorrectedTimestamp() {
         long currentTimestamp = System.currentTimeMillis();
-        if (currentTimestamp == previousTimestamp) {
+        if (currentTimestamp > previousTimestamp) {
+            final long oneYearInMillis = TimeUnit.DAYS.toMillis(365);
+            int maxCrazyNtpValueRetries = 1000;
+            while (currentTimestamp - previousTimestamp > oneYearInMillis) {
```
Comment (annym) [inline runtime/src/main/java/org/corfudb/runtime/view/CorfuGuidGenerator.java:240]: should this sleep in between retries? how do we know this is "enough" cycles/time to typically correct itself?

Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/view/CorfuGuidGenerator.java:240]: just experimental observations at this point - when ntp throws time around, it ends up correcting itself in subsequent calls. I just thought adding a sleep might slow down a problem that corrects itself quickly?

File: runtime/src/main/java/org/corfudb/runtime/view/CorfuGuidGenerator.java (hunk 1 above)
Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/view/CorfuGuidGenerator.java:258]: they are different right - currentTime > previousTimestamp means normal time moving forward, we only need to check for crazy jumps,
currentTime == previousTime means this call was invoked within the same millisecond resolution
But if currentTime < previousTime means time has moved backward and monotonic growth is broken so we fix up by adding drift correction.

Comment (annym) [reply runtime/src/main/java/org/corfudb/runtime/view/CorfuGuidGenerator.java:258]: correct, but if it is not greater and it is not equal, it necessarily is lower, so an else would be enough, right? (without the if)

File: runtime/src/main/java/org/corfudb/runtime/view/CorfuGuidGenerator.java (hunk 2 above)
Comment (annym) [reply runtime/src/main/java/org/corfudb/runtime/view/CorfuGuidGenerator.java:240]: ok. I was just wondering based on what we knew this was guaranteed to be enough time to recover... and one more question, what are the implications of this happening? is a warning enough?

Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/view/CorfuGuidGenerator.java:240]: This loop will most likely execute way faster than any clock correction. I don't see how it addresses the time jump issue.

**Review by Maithem — COMMENTED** [review]
Is this GUID generator still necessary for the queue?

Since the queue relies on the sequencer clock to generate ids can this generator be removed?
With the write pre-hook (using the sequencer result to populate the write payload) I don't see the importance of this GUID generator. The sequencer clock is already monotonic.

File: runtime/src/main/java/org/corfudb/runtime/view/CorfuGuidGenerator.java (hunk 1 above)
Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/view/CorfuGuidGenerator.java:258]: oh, got it, sure, I can remove the redundant if

File: runtime/src/main/java/org/corfudb/runtime/view/CorfuGuidGenerator.java (hunk 2 above)
Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/view/CorfuGuidGenerator.java:240]: logs had showed that in one call to currentTimeMillis() the value was something like INT_MAX and in the next call it had returned back to normal. I don't understand NTP to know why that happens, I can add a sleep if you think that helps.

**Comment by hisundar** [comment]
> Is this GUID generator still necessary for the queue?
> 
> Since the queue relies on the sequencer clock to generate ids can this generator be removed?
> With the write pre-hook (using the sequencer result to populate the write payload) I don't see the importance of this GUID generator. The sequencer clock is already monotonic.

GUID is still needed for the key, The pre-hook value is not used as key, it is used in the metadata section for storing the order.

File: runtime/src/main/java/org/corfudb/runtime/view/CorfuGuidGenerator.java (hunk 2 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/view/CorfuGuidGenerator.java:240]: I don't feel like spinning on currentTimeMillis is robust. Why isn't it sufficient to just retry the whole transaction from a higher layer? All transactions are susceptible to non-conflict related aborts and need to retry anyways.

Comment (hisundar) [reply runtime/src/main/java/org/corfudb/runtime/view/CorfuGuidGenerator.java:240]: we don't have control over that aspect right - I agree in general with the fact that they are supposed to retry the transaction, but according to them this part of the code should never throw any exception

**Comment by hisundar** [comment]
This isn't the right fix for the NTP issue, abandoning after discussion.

**Pull request closed without merging**